### PR TITLE
Handle mesh-provider IBC Failure Cases

### LINKED ISSUE #37: Implement mesh-provider IBC packet failure cases
See `contracts/mesh-provider/src/ibc.rs`:
- [ ] `fail_list_validators`: send another ListValidators message
- [ ] `fail_stake`:  release the bonded stake, adjust number
- [ ] `fail_unstake`: unrelease the bonded stake, remove claim (see [comment](https://github.com/CosmWasm/mesh-security/blob/main/contracts/mesh-provider/src/ibc.rs#L242))

### PR DESCRIPTION
**Pull request by teddyknox**
EDIT 10/19

I went through and implemented some todos related to IBC failures in mesh-provider. The only unintuitive change involves a shift from unilateral staking to bilateral staking. That is, mesh-provider only updates its staking state after it receives confirmation via ack from mesh-consumer on the "stake" IBC call. By contrast, mesh-provider continues to offer unilateral unstaking, since we must assume that the consumer chain might be byzantine.

## Qs
* @ethanfrey is there a way to mock IBC calls for testing?

Closes #37

### CODE REVIEW

**Comment by JakeHartnell** [comment]
> How should the mesh-provider ibc handler notify the UI that the stake/unstake operation failed? My intuition is that we'd want to publish an event...?

An event sounds reasonable to me.

> After the list_validators IBC call has been attempted 5 times, we give up. Would it be desirable to offer an execution endpoint for restarting the retries?

Perhaps? Can probably hold off on this for now. It's easy enough to add later. Maybe @ethanfrey has different thoughts.

> Should the retry count for list_validators be configurable or is a constant OK?

IMO constant is OK for now, but if you want to make it configurable we can.

File: contracts/mesh-provider/src/ibc.rs [hunk 1]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    // revert state changes associated with stake
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED.load(deps.storage, (&staker, &validator))?;
+    stake.stake_validator(&mut val, amount);
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+
+    // TODO: How should the user be notified of the failure?
```
Comment (JakeHartnell) [inline contracts/mesh-provider/src/ibc.rs:330]: IMO, we just add the relevant events / attributes, and leave it to the frontend.

File: contracts/mesh-provider/src/ibc.rs [hunk 2]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    let cfg = CONFIG.load(deps.storage)?;
+    let ready = env.block.time.plus_seconds(cfg.unbonding_period);
+    CLAIMS.create_claim(
+        deps.storage,
+        &staker,
+        amount,
+        Expiration::AtTime(ready),
+    )?;
+
+    // TODO: How should the user be notified of the failure?
```
Comment (JakeHartnell) [inline contracts/mesh-provider/src/ibc.rs:304]: IMO, we just add the relevant events / attributes, and leave it to the frontend.

Comment (JakeHartnell) [inline contracts/mesh-provider/src/ibc.rs:304]:
```suggestion
```

File: contracts/mesh-provider/src/contract.rs [hunk 3]
```diff
@@ -192,16 +194,6 @@ pub fn execute_unstake(
...
-    // create a future claim on number of shares (so we can adjust for later slashing)
-    let cfg = CONFIG.load(deps.storage)?;
-    let ready = env.block.time.plus_seconds(cfg.unbonding_period);
-    CLAIMS.create_claim(
-        deps.storage,
-        &info.sender,
-        amount,
-        Expiration::AtTime(ready),
-    )?;
-
```
Comment (JakeHartnell) [inline contracts/mesh-provider/src/contract.rs:204]: Cool, I think it makes sense to move this to the ack.

**Comment by ethanfrey** [comment]
I agree with @JakeHartnell comments on how it should work. Going to review the code now.

File: contracts/mesh-provider/src/contract.rs [hunk 4]
```diff
@@ -37,6 +38,7 @@ pub fn instantiate(
         slasher: None,
         lockup: deps.api.addr_validate(&msg.lockup)?,
         unbonding_period: msg.unbonding_period,
+        list_validators_retries_remaining: LIST_VALIDATORS_RETRIES,
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/contract.rs:41]: Part of me would store this in it's own Item, as the rest of those change very rarely and only on admin message.
Kind of "separation of concerns" (but this is a nit, open to counterarguments)

File: contracts/mesh-provider/src/contract.rs [hunk 5]
```diff
@@ -192,16 +194,6 @@ pub fn execute_unstake(
     STAKED.save(deps.storage, (&info.sender, &validator), &stake)?;
     VALIDATORS.save(deps.storage, &validator, &val)?;
 
-    // create a future claim on number of shares (so we can adjust for later slashing)
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/contract.rs:195]: Mental model from Sunny: "The provider chain should be resilient in face of Byzantine action from the consumer chain".

This means, the *provider* enforces the unstaking/unbonding period, and only slashes if it receives proof. It doesn't rely on the consumer chain to release tokens or slash arbitrarily.

Following this model, we should always unstake on the provider. We just want to ensure it makes it to the consumer if the consumer if honest. Staking on the other hand, should only happen if the consumer accepts it.

(Basically, we only lock up your tokens if the consumer promises to honor them. But we unlock them according to our local rules, just informing the consumer of the action).

File: contracts/mesh-provider/src/ibc.rs [hunk 6]
```diff
@@ -172,6 +174,14 @@ pub fn ibc_packet_ack(
             },
             false,
         ) => fail_stake(deps, key, validator, amount),
+        (
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:177]: Again, this covers failure mode for unstake. What you did for unstake, you should really do for stake instead.

And for unstake, something like you do for list_validators.

File: contracts/mesh-provider/src/ibc.rs [hunk 7]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
     for val in res.validators {
         VALIDATORS.save(deps.storage, &val, &Validator::new())?;
     }
+
+    // reset retry counter
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:227]: :+1: good to reset here

File: contracts/mesh-provider/src/ibc.rs [hunk 8]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
 
-pub fn fail_list_validators(_deps: DepsMut) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: send another ListValidators message
-    unimplemented!();
+pub fn fail_list_validators(deps: DepsMut, env: Env) -> Result<IbcBasicResponse, ContractError> {
+    // check if we should retry
+    let cfg = CONFIG.load(deps.storage)?;
+    let channel_id = CHANNEL.load(deps.storage)?;
+    if cfg.list_validators_retries_remaining <= 0 {
+        return Err(ContractError::NoResponse("list_validators".to_string(), channel_id))
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:240]: An error here (in ack) will abort the receipt of the acknowledgement packet. This is usually not what we want to do.
We can let it come and record this as a failure somehow. Emitting a custom event is a good idea.

I would not just `add_attribute`, but rather `add_event` with a custom type just for these failure modes, for easy searching in the UI.

File: contracts/mesh-provider/src/ibc.rs [hunk 9]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    }
+    let mut cfg_after = cfg.clone();
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:242]: I see no reason to clone. I would just declare `let mut cfg = CONFIG.load()` when it is created

File: contracts/mesh-provider/src/ibc.rs [hunk 10]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    cfg_after.list_validators_retries_remaining -= 1;
+    CONFIG.save(deps.storage, &cfg_after)?;
+
+    // retry
+    let packet = ProviderMsg::ListValidators {};
+    let msg = IbcMsg::SendPacket {
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:248]: looks good

File: contracts/mesh-provider/src/ibc.rs [hunk 11]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:266]: Again, great to error on user facing calls. Best not to in IBC message handling (as it will cause errors on the relayer, no events to the user). Early return would be good `return Ok(Response::new());`

File: contracts/mesh-provider/src/ibc.rs [hunk 12]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:271]: Maybe we return Response with a custom event saying this?

File: contracts/mesh-provider/src/ibc.rs [hunk 13]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
+    let mut stake = STAKED
+        .may_load(deps.storage, (&staker, &validator))?
+        .unwrap_or_default();
+    stake.unstake_validator(&mut val, amount)?;
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:275]: This will force the full unbonding period.
But since the stake was never acknowledged on the consumer side, we shouldn't do that.

I would not lock the stake in the ExecuteMsg, and rather lock it in the `ack_stake` handler. `fail_stake` handler doesn't have any staking/unbonding to deal with and rather just needs to call `release_claim` on the `mesh-lockup` contract (what happens now on claim, after the full unbonding period has passed)

File: contracts/mesh-provider/src/ibc.rs (hunk 2 above)
Comment (ethanfrey) [reply contracts/mesh-provider/src/ibc.rs:304]: Let's just add some event to the Responses here (and make them successes).

How about defining some standard event type to hold all errors (it is just arbitrary keys, but spec'ing it in the readme helps). Then something like:

```rust
let evt = Event::new("mesh_ibc_error")
  .add_attribute("action", "unstake")
  .add_attribute("error", err);
```

(Where `err` is any `ContractError` created earlier in the handling)

File: contracts/mesh-provider/src/ibc.rs [hunk 14]
```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
...
-    // TODO: unrelease the bonded stake, remove claim
-    // Maybe we only make Claim on ack?
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // revert state changes associated with stake
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:321]: Wait. You no longer set the unstake in the ExecuteMsg, but only in `ack_unstake`. Then this failure wouldn't need to hold anything.

This comment you can translate to the "stake" changes. For unstake, as I mentioned above, we just do it in the ExecuteMsg handler, ack is a no-op, and fail emits an event (maybe retrying list ListValidators)

Comment (ethanfrey) [reply contracts/mesh-provider/src/ibc.rs:321]: I actually think handling unexpected failures (that should only occur on error in the consumer side, not due to possible bad arguments from the provider). I mean, retry was my first idea, but does it make sense?

Should we flag the channel as "messed up" and freeze stuff til some gov account does a reset.

This is totally out of scope of this PR and I am happy with the retry in fail_list_validators, but let's not add more retry now and think a bit larger on how to handle such situations.

**Review by ethanfrey — COMMENTED** [review]
Made comments that handling of unstake should actually be like it was done. But stake is the one that should have such changes. And a bunch of minor points

File: contracts/mesh-provider/src/ibc.rs (hunk 14 above)
Comment (teddyknox) [reply contracts/mesh-provider/src/ibc.rs:321]: OK

**Review by teddyknox — COMMENTED** [review]
Thanks for the review Ethan. I've pushed with edits.

**Review by JakeHartnell — APPROVED** [review]
LGTM

**Review by Art3miX — COMMENTED** [review]
I'm not sure what should be covered under the POC, but `RETRIES` storage is shared "state", so if for example I stake and fail, the next successful stake call will reset `RETRIES`, or another example, if I stake and fail and you unstake and fail, we both will have 5 retries instead of 5 per each.

I think (and I might be wrong) that we should move retries per case (stake, unstake, all_val) into its own storages, and also use an ID, which can be the sender addr, so we can only check retries for that sender.

Also a question regarding the above, what should we do in the example:
"I stake and fail and next msg I unstake and fail."
I should not be able to unstake when im still retrying to stake.

Maybe we should keep stake and unstake retries shared under 1 storage, and when we call a msg, we first check if we are in a process of retrying, if we are still retrying we just fail the TX with "Still retrying last messages" sort of messages.
Basically allow to move on if we are not retrying.

**Comment by ethanfrey** [comment]
Yeah, I mentioned in some comment that retries make sense for the ListValidators query, which is required for startup, to avoid any hickups on startup. This is valid global state - should only be called once.

But we should just handle stake and unstake failures locally. Stake by reverting the deposit (as if they never did anything, no unbonding period). Unstake by just ignoring failure for MVP and open an issue how to better communicate it (point is byzantine network should not be able to prevent honest provider from unstaking)

File: contracts/mesh-provider/src/ibc.rs [hunk 15]
```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
...
+}
+
+fn ack_stake(deps: DepsMut, staker: String, validator: String, amount: Uint128) -> Result<IbcBasicResponse, ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Ok(IbcBasicResponse::new().add_event(Event::new("ack_stake_zero_amount")));
+    }
+
+    let mut val = VALIDATORS
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:262]: This is basically the code from https://github.com/CosmWasm/mesh-security/pull/38/files#diff-ba0374507dc354d977b251871cc89ff4552166c72b7b326edb14a9686ce8d5bdL118-L128 which is good.

I think the below point was not properly resolved (returning an event rather than erroring in ack), but this case should never happen.

File: contracts/mesh-provider/src/ibc.rs [hunk 16]
```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
...
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED.load(deps.storage, (&staker, &validator))?;
+    stake.stake_validator(&mut val, amount);
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:269]: Remove retries.

File: contracts/mesh-provider/src/ibc.rs [hunk 17]
```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
...
-    _amount: Uint128,
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let mut retries = RETRIES.load(deps.storage)?;
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:285]: Remove all retries logic here. Only on list_validators.

What we **do** need to do is actually return the stake to the user. We never updated our internal stake/validator book-keeping (that is done on `ack_stake`), but we did receive some bonded tokens from the lockup contract. We should tell that contract that this user will not be bonding this stake and can release it there. Otherwise, those tokens remain locked forever in failure mode.

(Look at logic called on successful claim, after unbonding, to see how to release it)

File: contracts/mesh-provider/src/contract.rs [hunk 18]
```diff
@@ -115,17 +120,7 @@ pub fn execute_receive_claim(
         return Err(ContractError::ZeroAmount);
     }
 
-    let mut val = VALIDATORS
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/contract.rs:118]: We actually probably want this check here, even if we don't use the result.

Erroring quickly when they try to bond to random address that doesn't represent a validator, rather than handling that in `fail_stake` is much better UX.

File: contracts/mesh-provider/src/ibc.rs [hunk 19]
```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
...
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: unrelease the bonded stake, remove claim
```
Comment (ethanfrey) [inline contracts/mesh-provider/src/ibc.rs:242]: I would make this a no-op.
No retries, no action needed.
Same as `ack_unstake`

**Review by ethanfrey — COMMENTED** [review]
Main point is removing retries from stake and unstake. I feel rather strong about stake - we try once and undo if it fails.

For unstake (where we don't take actions based on results), I see no harm in retrying, but it does seem unnecessary and if done properly will need a per-packet/sender counter, which is more complex.

**Review by JakeHartnell — CHANGES_REQUESTED** [review]
Please rebase and implement the changes Ethan has suggested.

File: contracts/mesh-provider/src/contract.rs (hunk 18 above)
Comment (teddyknox) [reply contracts/mesh-provider/src/contract.rs:131]: Good point, will add it back.

File: contracts/mesh-provider/src/ibc.rs [hunk 20]
```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
     for val in res.validators {
         VALIDATORS.save(deps.storage, &val, &Validator::new())?;
     }
-    Ok(IbcBasicResponse::new())
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply contracts/mesh-provider/src/ibc.rs:255]: Interesting. Thanks for the tip.

File: contracts/mesh-provider/src/ibc.rs (hunk 17 above)
Comment (teddyknox) [reply contracts/mesh-provider/src/ibc.rs:312]: Sorry for my absence, I've been onboarding to a new job. Is it safe to assume we should be returning back 
`coin(amount.u128(), config.rewards_ibc_denom)`?
